### PR TITLE
CIT-41 #time implement PATCH,PUT,DELETE,POST method to perform user s…

### PR DESCRIPTION
**Pull request by codal-uchoudhary**
…pecific CRUD operations in the Blog app

### CODE REVIEW

**Comment by codal-uchoudhary** [comment]
In the Blog app
1 Only Authenticate user can created Blogs
2 Valid Author can delete and Update his/her blogs 
3 Author can fetch his/her blogs

**Comment by codal-uchoudhary** [comment]
Screenshot 2024-02-25 221443

File: Projects/My_website/myWebsite/blog/views.py [hunk 1]
```diff
@@ -26,16 +26,58 @@ class userBlog(APIView):
     authentication_classes = [TokenAuthentication]
     permission_classes = [IsAuthenticated]
 
+    def get_item(self,slug):
+        item = Post.objects.get(slug=slug)
```
Comment (arvish-codal) [inline Projects/My_website/myWebsite/blog/views.py]: @codal-uchoudhary 
What if there's no data that that matches the slug name ? Won't the `GET` query throw an error.
Make necessary changes.

File: Projects/My_website/myWebsite/blog/views.py [hunk 2]
```diff
@@ -26,16 +26,58 @@ class userBlog(APIView):
...
+        serializers = blogSerializer(data=data)
+        if serializers.is_valid():
+            serializers.save()
+            return Response(serializers.data)
+        return Response(serializers.errors)
+    
+    def put(self,request,slug):
+        data = request.data
+        if data['author'] != request.user.id:
+            return Response({'message':"you are not valid author"})
```
Comment (arvish-codal) [inline Projects/My_website/myWebsite/blog/views.py]: @codal-uchoudhary , what are we doing here ? 
You need to check whether the blog that is being edited is of the same user that is currently logged in.

we do not need to pass author key in the EDIT request payload. 

same mistake is in the patch & delete as well.

File: Projects/My_website/myWebsite/blog/views.py [hunk 3]
```diff
@@ -26,16 +26,58 @@ class userBlog(APIView):
...
+        serializer = blogSerializer(obj,data=data)
+        if serializer.is_valid():
+            serializer.save()
+            return Response(serializer.data)
+        return Response(serializer.errors)
+    
+    def patch(self,request,slug):   
+        data = request.data
+        if data['author'] != request.user.id:
+            return Response({'message':"you are not valid author"})
```
Comment (arvish-codal) [inline Projects/My_website/myWebsite/blog/views.py]: SUGGESTION : Pls use more generic messages here. Like `You are not authorised to perform this action.`

File: Projects/My_website/myWebsite/blog/views.py (hunk 1 above)
Comment (codal-uchoudhary) [reply Projects/My_website/myWebsite/blog/views.py]: > @codal-uchoudhary What if there's no data that that matches the slug name ? Won't the `GET` query throw an error. Make necessary changes.

Updated

Comment (codal-uchoudhary) [reply Projects/My_website/myWebsite/blog/views.py:30]: > @codal-uchoudhary What if there's no data that that matches the slug name ? Won't the `GET` query throw an error. Make necessary changes.

Comment (codal-uchoudhary) [reply Projects/My_website/myWebsite/blog/views.py]: Screenshot 2024-02-26 114400

Comment (arvish-codal) [reply Projects/My_website/myWebsite/blog/views.py]: @codal-uchoudhary 
In the delete request, why are we passing `Money-uhdbfu-dsfjn` ? What & why is that ? To delete a blog, don't 
you think we must be passing ID as the unique identifier ?

File: Projects/My_website/myWebsite/blog/views.py (hunk 2 above)
Comment (codal-uchoudhary) [reply Projects/My_website/myWebsite/blog/views.py]: @arvish-codal  , here i am comparing the Author id of the blog  with current user id . to check the author is valid or not.

Comment (codal-uchoudhary) [reply Projects/My_website/myWebsite/blog/views.py]: if it return True then , the author is the right person to edit this blog.

Comment (codal-uchoudhary) [reply Projects/My_website/myWebsite/blog/views.py]: I tested PUT,PATCH,DELETE APIs in the postman .

Comment (arvish-codal) [reply Projects/My_website/myWebsite/blog/views.py]: @codal-uchoudhary , yeah correct, that's what we need to do. But i think you're also passing author key in the payload. We should not be doing that.

the flow should be :
  --> there should be blog ID in the PUT request endpoint : example /blog/{id} --> /blog/1
  --> now from the Blog ID, you should be fetching the Blog instance.
  --> From the received blog instance, compare the blog's author ID with the request user id. 
  
 This is how we AUTHORISE whether the blog being edited / deleted is posted by the logged in user or not.

File: Projects/My_website/myWebsite/blog/views.py (hunk 1 above)
Comment (codal-uchoudhary) [reply Projects/My_website/myWebsite/blog/views.py]: @arvish-codal  , my intension is to pass wrong slug to check error , we get invalid slug in the response

Comment (arvish-codal) [reply Projects/My_website/myWebsite/blog/views.py]: @codal-uchoudhary , in that case. Pass any random blog ID..i.e the one which doesn't exist in the DB at all.

**_

> Unique identified for a blog should always be it's ID...no any other field.

_**

File: Projects/My_website/myWebsite/blog/views.py (hunk 2 above)
Comment (codal-uchoudhary) [reply Projects/My_website/myWebsite/blog/views.py]: okay ! let me change the methods

Comment (codal-uchoudhary) [reply Projects/My_website/myWebsite/blog/views.py]: @arvish-codal  , changed as per comment

File: Projects/My_website/myWebsite/blog/views.py [hunk 4]
```diff
@@ -26,16 +26,57 @@ class userBlog(APIView):
     authentication_classes = [TokenAuthentication]
     permission_classes = [IsAuthenticated]
 
+    def get_item(self,slug):
+        try:
+            item = Post.objects.get(id=slug)
+            return item
+        except:
```
Comment (arvish-codal) [inline Projects/My_website/myWebsite/blog/views.py]: @codal-uchoudhary ,
Pls explain how can `id=slug` condition be used to fetch the Post instance ?

Comment (codal-uchoudhary) [reply Projects/My_website/myWebsite/blog/views.py]: so , now i add id of post in the URL

Comment (codal-uchoudhary) [reply Projects/My_website/myWebsite/blog/views.py]: image

**Review by arvish-codal — APPROVED** [review]

**Pull request merged**